nat-lab/guidelines.md

```diff
@@ -0,0 +1,215 @@
+### Natlab Guidelines:
+
+1. **Use Parametrization:**
+   - Test across all operating systems when possible.
+   - Ensure parametrization for each node.
+
+   **Why:**
+   - Covers more areas.
+   - Better readability.
+   - Wider usage of a test.
+
+   **Example:**
+   ```
+   @pytest.mark.parametrize("alpha_node", [ConnectionTag.MAC_VM, ConnectionTag.WINDOWS_VM, ConnectionTag.DOCKER_CONE_CLIENT_1])
+   @pytest.mark.parametrize("beta_node", ConnectionTag.MAC_VM)
+   def test_params(alpha_node, beta_node):
+   ```
+
+2. **Assertions Clarity:**
+   - Assertions must be explicit and easily visible in the test.
+   - Avoid assertions in different threads, keep it within test, raise an exception instead if needed.
+
+   **Why:**
+   - Easier to track which test asserted if assert is within a test.
+   - Easier to read stacktrace.
+
+3. **Start Event Checks Promptly:**
+   - Initiate event checks before the corresponding action begins.
+
+   **Why:**
+   - Starting to wait for event after action, might result in missing the event.
+
+   **Example**:
+   - DON'T:
+   ```
+      await connect_to_vpn()
+      # event might occur event before starting to wait
+      await wait_for_event_peer(State.Connected)
+   ```
+   - DO:
+   ```
+      async with asyncio_util.run_async_context(wait_for_event_peer(State.Connected)) as event:
+            await asyncio.wait_for(
+                asyncio.gather(
+                    *[
+                        connect_to_vpn()
+                        event,
+                    ]
+                ),
+                timeout
+            )
+   ```
+
+4. **Utilize Provided Helpers:**
+   - Employ provided helper functions to establish the test environment.
+
+5. **Timeouts for Potential Hangs:**
+   - Apply timeouts to actions susceptible to hanging.
```

Comment:
I don't think this is actionable - how do I decide if something is "susceptible to hanging"?

Reply:
You wouldn't know all of the places, but for those, like in the example, you can add timeouts.

Reply:
Currently this is (imho) almost like "don't make mistakes". First of all we don't have per test timeout so you can't rely on overall test timeout to detect a failure (right, or am I mistaken?). Second, we often expect to reach certain states in specified time so adding a timeout can be used to verify that we are within the required time. Finally, adding timeout to commands that perform network transmissions (eg. nslookup) might also be a good idea since we expect networks to be sometimes slow/broken.

In general adding some sort of heuristic of at least few distinct examples will help people to understand when this is good idea.

Reply:
1. We have timeout
2. Maybe I should just rephrase `potential hangs` to `wanted state not being reached`

Reply:
I think that few distinct examples of situations when to add timeout would be useful, but maybe I'm overthinking this and this is fine as it is. Consider this a nitpick and feel free to skip it :)